scripts/make_luals_meta.py

```diff
@@ -58,116 +62,272 @@
+                lines.append(stripped)
+            else:
+                lines[-1] += "\n--- " + stripped
+
+        return lines if lines else None
+
+    def read_class_body(self) -> list[list[str]]:
+        """The reader must be at the first line of the class/struct body. All comments inside the class are returned."""
+        items = []
+        while (line := self.peek_line()) is not None:
+            if line.startswith("};"):
+                self.next_line()
+                break
+            if not is_comment_start(line):
+                self.next_line()
+                continue
+            doc = self.next_doc_comment()
+            if not doc:
+                break
+            items.append(doc)
+        return items
+
+    def read_enum_variants(self) -> list[str]:
+        """The reader must be before an enum class definition (possibly with some comments before). It returns all variants."""
+        items = []
+        is_comment = False
+        while (line := self.peek_line()) is not None and not line.startswith("};"):
+            self.next_line()
+            if is_comment:
+                if line.endswith("*/"):
+                    is_comment = False
+                continue
+            if line.startswith("/*"):
+                is_comment = True
+                continue
+            if line.startswith("//"):
+                continue
+            if line.endswith("};"):  # oneline declaration
+                opener = line.find("{") + 1
+                closer = line.find("}")
+                items = [
+                    line.split("=", 1)[0].strip()
+                    for line in line[opener:closer].split(",")
+                ]
+                break
+            if line.startswith("enum class"):
+                continue
+
+            items.append(line.rstrip(","))
+
+        return items
+
+
+def finish_class(out, name):
+    out.write(f"{name} = {{}}\n")
+
+
+def printmsg(path: Path, line: int, message: str):
+    print(f"{path.relative_to(repo_root)}:{line} {message}")
+
+
+def panic(path: Path, line: int, message: str):
+    printmsg(path, line, message)
+    exit(1)
+
+
+def write_func(path: Path, line: int, comments: list[str], out: TextIOWrapper):
+    if not comments[0].startswith("@"):
+        out.write(f"--- {comments[0]}\n---\n")
+        comments = comments[1:]
+    params = []
+    for comment in comments[:-1]:
+        if not comment.startswith("@lua"):
+            panic(path, line, f"Invalid function specification - got '{comment}'")
+        if comment.startswith("@lua@param"):
+            params.append(comment.split(" ", 2)[1])
+
+        out.write(f"---{comment.removeprefix('@lua')}\n")
+
+    if not comments[-1].startswith("@exposed "):
+        panic(path, line, f"Invalid function exposure - got '{comments[-1]}'")
+    name = comments[-1].split(" ", 1)[1]
+    printmsg(path, line, f"function {name}")
+    out.write(f"function {name}({", ".join(params)}) end\n\n")
```

Comment:
This requires Python 3.12, use different quotes or extract the join call into a variable. Also my syntax highlighter doesn't like it

```suggestion
    out.write(f"function {name}({', '.join(params)}) end\n\n")
```